library/ssl_tls.c

```diff
@@ -4889,7 +4889,35 @@ const char *mbedtls_ssl_get_version( const mbedtls_ssl_context *ssl )
 }
 
 #if defined(MBEDTLS_SSL_MAX_FRAGMENT_LENGTH)
-size_t mbedtls_ssl_get_max_frag_len( const mbedtls_ssl_context *ssl )
+size_t mbedtls_ssl_get_input_max_frag_len( const mbedtls_ssl_context *ssl )
+{
+    size_t max_len = MBEDTLS_SSL_MAX_CONTENT_LEN;
+
+    /* Use the configured MFL for the client if we're past SERVER_HELLO_DONE */
+    if( ssl->conf->endpoint == MBEDTLS_SSL_IS_CLIENT &&
+        ssl->state >= MBEDTLS_SSL_SERVER_HELLO_DONE )
+    {
+        return ssl_mfl_code_to_length( ssl->conf->mfl_code );
+    }
+
+    /* Check if a smaller max length was negotiated */
+    if( ssl->session_out != NULL &&
+        ssl_mfl_code_to_length( ssl->session_out->mfl_code ) < max_len )
+    {
+        max_len = ssl_mfl_code_to_length( ssl->session_out->mfl_code );
```

Comment:
This call `ssl_mfl_code_to_length( ssl->session_out->mfl_code );` is duplicated and I'm not sure if different compilers will be able to optimize it themselves. I would recommend to use one more value and do only one call of this function

Reply:
This function doesn't compute anything, only returns a value based on a switch, so it's a minor optimization. I'm not sure if the code is more readable this way either:
```suggestion
if( ssl->session_out != NULL &&
    ( read_mfl = ssl_mfl_code_to_length( ssl->session_out->mfl_code ) )
      < max_len )
    {
        max_len = read_mfl;
    }
```

Reply:
Yes, I also think it is not more readable... buy you can do it that:
```
if( ssl->session_out != NULL )
{
    read_mfl = ssl_mfl_code_to_length( ssl->session_out->mfl_code );
    if( max_len < max_len )
    {
        max_len = read_mfl;
    }
}
```
In for a penny, in for a pound ;).